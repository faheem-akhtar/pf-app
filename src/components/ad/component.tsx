import { FunctionComponent } from 'react';

import styles from './ad.module.scss';

export const AdComponent: FunctionComponent<{ id: string }> = ({ id }) => <div data-ad className={styles.ad} id={id} />;
