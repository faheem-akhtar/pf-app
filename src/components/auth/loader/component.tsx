import { domClassMerge } from 'helpers/dom/class-merge';

import styles from './auth-loader.module.scss';

export const AuthLoaderComponent = ({
  isEnabled,
  isCentered,
}: {
  isEnabled: boolean;
  isCentered: boolean;
}): JSX.Element | null =>
  isEnabled ? (
    <div className={styles.loader1}>
      <div
        className={domClassMerge(styles['loader1__spinner'], isCentered ? styles['loader1__spinner--centered'] : '')}
      >
        <div className={domClassMerge(styles['loader1__bounce'], styles['loader1__bounce--1'])} />
        <div className={domClassMerge(styles['loader1__bounce'], styles['loader1__bounce--2'])} />
        <div className={domClassMerge(styles['loader1__bounce'], styles['loader1__bounce--3'])} />
      </div>
    </div>
  ) : null;
