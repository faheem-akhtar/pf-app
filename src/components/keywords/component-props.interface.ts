import { FiltersCategoryIdEnum } from 'enums/filters/category-id.enum';

export interface KeywordsComponentPropsInterface {
  /**
   * On change
   */
  onChange: (value: string) => void;
  /**
   * Value
   */
  value: string;
  /**
   * Current selected category
   */
  category: FiltersCategoryIdEnum;
  /**
   * If true, the chips will be rendered on the bottom of the input, otherwise inside the input
   */
  chipsOnTheBottom?: true;
  /**
   * additional css classes for component
   */
  className?: string;
  /**
   * additional css classes for chips
   */
  chipClassName?: string;
  /**
   * keywords-category mapping
   */
  keywordsMapping: Record<FiltersCategoryIdEnum, string[]>;
}
