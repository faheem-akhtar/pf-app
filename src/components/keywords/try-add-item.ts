import { MultiSelectionAutocompleteExtensionInterface } from 'library/multi-selection-autocomplete/extension.interface';

export const keywordsTryAddItem = (props: MultiSelectionAutocompleteExtensionInterface<string>): void => {
  const inputValueNoSpace = props.inputValue.replace(/\s/g, '');
  if (inputValueNoSpace.length) {
    props.addItem(props.inputValue);
  }
};
