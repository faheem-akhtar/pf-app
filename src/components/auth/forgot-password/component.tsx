import React, { useState } from 'react';

import { AuthForgotPasswordFieldEnum } from 'components/auth/forgot-password/field.enum';
import { AuthForgotPasswordPropsInterface } from 'components/auth/forgot-password/props.interface';
import { AuthLoaderComponent } from 'components/auth/loader/component';
import { AuthLoginFieldEnum } from 'components/auth/login/field.enum';
import { AuthResetPasswordService } from 'services/auth/reset-password.service';
import { ButtonComponentTypeEnum } from 'library/button/component-type.enum';
import { ButtonSizeEnum } from 'library/button/size.enum';
import { ButtonTemplate } from 'library/button/template';
import { domClassMerge } from 'helpers/dom/class-merge';
import { ErrorMessageComponent } from 'components/error-message/component';
import { formMakeValidator } from 'components/form/make-validator';
import { GoogleRecaptchaService } from 'services/google/recaptcha.service';
import { InputBaseComponent } from 'library/input/base/component';
import { ReCaptchaComponent } from 'components/re-captcha/component';
import { useTranslation } from 'helpers/translation/hook';
import { validationEmail } from 'helpers/validation/email';
import { validationRequired } from 'helpers/validation/required';

import styles from '../auth.module.scss';

export const AuthForgotPasswordComponent = (props: AuthForgotPasswordPropsInterface): JSX.Element => {
  const captchaService = GoogleRecaptchaService();
  const { t } = useTranslation();
  const [email, setEmail] = useState('');
  const [errors, setErrors] = useState<Partial<Record<AuthLoginFieldEnum, string>>>({});
  const [errorMessage, setErrorMessage] = useState('');
  const [validators] = useState(() => ({
    [AuthLoginFieldEnum.email]: [
      validationRequired(t('validation/empty-email')),
      validationEmail(t('validation/not-valid-email')),
    ],
  }));
  const [isLoading, setIsLoading] = useState(false);

  /**
   * Validate form fields
   */
  const validate = formMakeValidator(errors, setErrors, validators);

  // TODO-FE[CX-499] - Add success message
  return (
    <div data-testid='AuthForgotPasswordComponent'>
      <div className={styles.heading}>{t('auth/forgot-password')}?</div>
      <AuthLoaderComponent isEnabled={isLoading} isCentered={true} />
      {errorMessage && <ErrorMessageComponent message={errorMessage} />}
      <form
        onSubmit={(e): void => {
          e.preventDefault();

          if (
            !validate({
              [AuthForgotPasswordFieldEnum.email]: email,
            })
          ) {
            setIsLoading(true);
            captchaService.execute().then((captcha_token) => {
              AuthResetPasswordService({
                email,
                captcha_token,
              })
                .then((e) => {
                  if (e.ok) {
                    // Reset server error message
                    setErrorMessage('');
                    // Close modal
                    props.onClose();
                    props.onSuccess();
                  } else {
                    setErrorMessage(e.error?.body || `${t('auth/something-wrong')}! ${t('auth/try-later')}`);
                    captchaService.reset();
                  }
                })
                .finally(() => setIsLoading(false));
            });
          }
          return;
        }}
      >
        <div className={styles['input-area']}>
          <InputBaseComponent
            placeholder={t('email')}
            type='email'
            value={email}
            error={!!errors[AuthForgotPasswordFieldEnum.email]}
            onChange={(value): void => {
              validate({
                [AuthForgotPasswordFieldEnum.email]: value,
              });
              setEmail(value);
            }}
            errorText={errors[AuthForgotPasswordFieldEnum.email]}
          />
        </div>
        <div className={styles['input-area']}>
          <ButtonTemplate
            className={styles.button}
            type='submit'
            componentType={ButtonComponentTypeEnum.primary}
            size={ButtonSizeEnum.regular}
          >
            {t('auth/reset-password')}
          </ButtonTemplate>
        </div>
        <ReCaptchaComponent />
      </form>
      <div className={domClassMerge(styles.link, styles['create-account'])} onClick={props.onRegister}>
        {t('auth/not-registered-yet')}?&nbsp;{t('auth/create-account')}
      </div>
    </div>
  );
};
