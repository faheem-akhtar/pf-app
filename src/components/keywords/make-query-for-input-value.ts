import { keywordsByCategory } from 'config/keywords/by-category';
import { stringMakeCaseInsensitiveTester } from 'helpers/string/make-case-insensitive-tester';

import { FiltersCategoryIdEnum } from 'enums/filters/category-id.enum';

const MAX_NUMBER_OF_KEYWORDS = 8;

export const keywordsMakeQueryForInputValue =
  (category: FiltersCategoryIdEnum) =>
  (inputValue: string): Promise<string[]> => {
    const searchResults = [];
    const matchesInputValueCaseInsensitive = stringMakeCaseInsensitiveTester(inputValue);

    for (const keyword of keywordsByCategory[category]) {
      if (matchesInputValueCaseInsensitive(keyword)) {
        searchResults.push(keyword);
        if (searchResults.length === MAX_NUMBER_OF_KEYWORDS) {
          break;
        }
      }
    }

    return Promise.resolve(searchResults);
  };
