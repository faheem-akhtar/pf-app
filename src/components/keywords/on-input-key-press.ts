import { KeyboardKeyEnum } from 'enums/keyboard/key.enum';
import { MultiSelectionAutocompleteInputKeyPressExtensionPropsInterface } from 'library/multi-selection-autocomplete/input-key-press-extension-props.interface';

import { keywordsTryAddItem } from './try-add-item';

export const keywordsOnInputKeyPress = (
  props: MultiSelectionAutocompleteInputKeyPressExtensionPropsInterface<string>
): void => {
  if (props.e.key === KeyboardKeyEnum.enter) {
    keywordsTryAddItem(props);
    props.deactivate();
  }
  if (props.e.key === KeyboardKeyEnum.comma || props.e.key === KeyboardKeyEnum.commaArabic) {
    keywordsTryAddItem(props);
    props.clearInput();
    props.e.preventDefault();
  }
};
