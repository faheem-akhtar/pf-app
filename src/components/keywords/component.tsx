import { stringSanitizeSimple } from 'helpers/string/sanitize-simple';
import { useTranslation } from 'helpers/translation/hook';
import { MultiSelectionAutocompleteComponent } from 'library/multi-selection-autocomplete/component';

import { KeywordsComponentPropsInterface } from './component-props.interface';
import { keywordsMakeQueryForInputValue } from './make-query-for-input-value';
import { keywordsOnInputKeyPress } from './on-input-key-press';
import { keywordsTryAddItem } from './try-add-item';

const separator = ',';

export const KeywordsComponent = ({
  onChange,
  value,
  category,
  chipsOnTheBottom,
  className,
  chipClassName,
  keywordsMapping,
}: KeywordsComponentPropsInterface): JSX.Element => {
  const { t } = useTranslation();
  const keywords = value.split(separator).filter((x) => x);
  const keywordsQueryForInputValue = keywordsMakeQueryForInputValue(category, keywordsMapping);

  return (
    <MultiSelectionAutocompleteComponent
      className={className}
      chipClassName={chipClassName}
      chipsOnTheBottom={chipsOnTheBottom}
      value={keywords}
      getTitle={(keyword): string => keyword}
      getChipTitle={(keyword): string => keyword}
      placeholder={t('keywords_widget/placeholder')}
      onInputChange={keywordsQueryForInputValue}
      onFocus={keywordsQueryForInputValue}
      onAddItem={(keyword): void => {
        const trimmedWord = keyword.trim();
        const regex = new RegExp(trimmedWord, 'i');
        if (!keywords.some((word) => regex.test(word) && word.length === trimmedWord.length)) {
          onChange([...keywords, keyword].join(separator));
        }
      }}
      onItemRemove={(keyword): void => onChange(keywords.filter((k) => keyword !== k).join(separator))}
      onItemGroupRemoveClick={(): void => onChange(keywords[0])}
      renderNoSuggestions={(inputValue: string): JSX.Element => (
        <span
          // eslint-disable-next-line react/no-danger
          dangerouslySetInnerHTML={{
            __html: t('keywords_widget/no-suggestions-message', { inputValue: stringSanitizeSimple(inputValue) }),
          }}
        />
      )}
      onInputKeyPress={keywordsOnInputKeyPress}
      onWindowMouseDownOutsideAutocomplete={keywordsTryAddItem}
    />
  );
};
