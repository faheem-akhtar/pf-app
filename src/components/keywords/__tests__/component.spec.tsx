/**
 * @jest-environment jsdom
 */

import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { mockMiscAddTranslation } from 'mocks/misc/add-translation.mock';
import { configKeywordsByCategoryStub } from 'stubs/config/keywords/by-category.stub';

import { configKeywordsMaxWordLimit } from 'config/keywords/max-word-limit';
import { FiltersCategoryIdEnum } from 'enums/filters/category-id.enum';

import { KeywordsComponent } from '../component';

describe('KeywordsComponent', () => {
  let onChange: (value: string) => void;

  beforeEach(() => {
    onChange = jest.fn();
  });

  test('if render is completed without throwing any errors', () => {
    const { container } = render(
      <KeywordsComponent
        keywordsMapping={configKeywordsByCategoryStub}
        onChange={onChange}
        value='value'
        category={FiltersCategoryIdEnum.residentialForRent}
      />
    );
    expect(container).toMatchSnapshot();
  });

  test('if no-suggestion template rendered when suggestions are empty', async () => {
    mockMiscAddTranslation('keywords_widget/no-suggestions-message', 'no-message');
    render(
      <KeywordsComponent
        keywordsMapping={configKeywordsByCategoryStub}
        onChange={onChange}
        value=''
        category={FiltersCategoryIdEnum.residentialForRent}
      />
    );

    userEvent.tab();

    expect(screen.getByPlaceholderText('keywords_widget/placeholder')).toHaveFocus();
    userEvent.type(screen.getByPlaceholderText('keywords_widget/placeholder'), 'lorem');

    await waitFor(() => {
      expect(screen.getByText(/no-message|lorem/i)).toBeInTheDocument();
    });
  });

  it('should get the titles from the suggestions', async () => {
    render(
      <KeywordsComponent
        keywordsMapping={configKeywordsByCategoryStub}
        onChange={onChange}
        value=''
        category={FiltersCategoryIdEnum.residentialForRent}
      />
    );

    userEvent.tab();

    expect(screen.getByPlaceholderText('keywords_widget/placeholder')).toHaveFocus();

    userEvent.type(screen.getByPlaceholderText('keywords_widget/placeholder'), 'lor');

    const suggestionsTemplate = await screen.findByTestId('multi-selection-autocomplete-template-suggestions');
    expect(suggestionsTemplate.innerHTML).toContain(`<strong>lor</strong>`);
  });

  test('if suggestion chips are set correctly', async () => {
    const value = 'lorem,ipsum';

    render(
      <KeywordsComponent
        keywordsMapping={configKeywordsByCategoryStub}
        onChange={onChange}
        value={value}
        category={FiltersCategoryIdEnum.residentialForRent}
        chipsOnTheBottom
      />
    );

    await waitFor(async () => {
      const chipsContainerTemplate = await screen.findByTestId('multi-selection-autocomplete-template-chips');
      value.split(',').forEach((word) => {
        expect(chipsContainerTemplate.innerHTML).toContain(`<span class="chipLabel label">${word}</span>`);
      });
    });
  });

  test('if the given keyword is added to the existing keyword stack', async () => {
    const values = 'dolor,sit';
    render(
      <KeywordsComponent
        keywordsMapping={configKeywordsByCategoryStub}
        onChange={onChange}
        value={values}
        category={FiltersCategoryIdEnum.residentialForRent}
      />
    );

    userEvent.click(screen.getByTestId('multi-selection-autocomplete-template'));
    userEvent.tab();

    expect(screen.getByPlaceholderText('keywords_widget/placeholder')).toHaveFocus();

    userEvent.type(screen.getByPlaceholderText('keywords_widget/placeholder'), 'lor');

    const suggestionButton = await screen.findByTestId('multi-selection-autocomplete-template-suggestion-button');

    if (suggestionButton) {
      userEvent.click(suggestionButton);

      expect(onChange).toBeCalledWith(
        [...values.split(','), configKeywordsByCategoryStub[FiltersCategoryIdEnum.residentialForRent][0]].join(',')
      );
    }
  });

  test('if the given keyword is not added since it already exists', async () => {
    const values = 'lorem,ipsum';
    render(
      <KeywordsComponent
        keywordsMapping={configKeywordsByCategoryStub}
        onChange={onChange}
        value={values}
        category={FiltersCategoryIdEnum.residentialForRent}
      />
    );

    userEvent.click(screen.getByTestId('multi-selection-autocomplete-template'));
    userEvent.tab();

    expect(screen.getByPlaceholderText('keywords_widget/placeholder')).toHaveFocus();

    userEvent.type(screen.getByPlaceholderText('keywords_widget/placeholder'), 'lor');

    const suggestionButton = await screen.findByTestId('multi-selection-autocomplete-template-suggestion-button');
    if (suggestionButton) {
      userEvent.click(suggestionButton);
      expect(onChange).not.toHaveBeenCalledTimes(1);
    }
  });

  test('if the given keyword is not added since stack is full', async () => {
    const values = 'lorem,ipsum';
    render(
      <KeywordsComponent
        keywordsMapping={configKeywordsByCategoryStub}
        onChange={onChange}
        value={values}
        category={FiltersCategoryIdEnum.commercialForRent}
      />
    );

    userEvent.click(screen.getByTestId('multi-selection-autocomplete-template'));
    userEvent.tab();

    expect(screen.getByPlaceholderText('keywords_widget/placeholder')).toHaveFocus();

    userEvent.type(screen.getByPlaceholderText('keywords_widget/placeholder'), 'a');

    const suggestionButtons = await screen.findAllByTestId('multi-selection-autocomplete-template-suggestion-button');
    expect(suggestionButtons).toHaveLength(configKeywordsMaxWordLimit);
  });

  test('if the clicked keyword is removed from the keyword stack', () => {
    const values = 'dolor,sit';
    render(
      <KeywordsComponent
        keywordsMapping={configKeywordsByCategoryStub}
        onChange={onChange}
        value={values}
        category={FiltersCategoryIdEnum.residentialForRent}
        chipsOnTheBottom
      />
    );

    // first suggestion will be clicked
    const chipButton = screen.getAllByRole('button')[0];

    if (chipButton) {
      userEvent.click(chipButton);
      const [, lastSuggestion] = values.split(',');
      expect(onChange).toHaveBeenCalledWith(lastSuggestion);
    }
  });

  test('if the item group is removed when clicked', () => {
    const values = 'dolor,sit,amet,lorem';
    render(
      <KeywordsComponent
        keywordsMapping={configKeywordsByCategoryStub}
        onChange={onChange}
        value={values}
        category={FiltersCategoryIdEnum.residentialForRent}
      />
    );

    const chipButtons = screen.getAllByRole('button');

    if (chipButtons) {
      userEvent.click(chipButtons[1]);
      expect(onChange).toHaveBeenCalledWith(values.split(',')[0]);
    }
  });
});
