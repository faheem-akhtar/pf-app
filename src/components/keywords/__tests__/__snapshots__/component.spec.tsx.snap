// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KeywordsComponent if render is completed without throwing any errors 1`] = `
<div>
  <div
    class="root"
    data-testid="multi-selection-autocomplete-template"
  >
    <div
      class="container container_not_active_single_element"
    >
      <label
        class="chip container chip_selected"
      >
        <span
          class="chipLabel label"
        >
          value
        </span>
        <span
          class="chipSuffix"
        >
          <button
            class="cross_button"
          >
            <svg
              class="cross_icon"
              viewBox="3 3 10 10"
            >
              <path
                clip-rule="evenodd"
                d="M4.34408 3.17475C4.01717 2.92057 3.54448 2.94367 3.24408 3.24408C2.91864 3.56951 2.91864 4.09715 3.24408 4.42259L6.82149 8L3.24408 11.5774C2.91864 11.9028 2.91864 12.4305 3.24408 12.7559C3.54448 13.0563 4.01717 13.0794 4.34408 12.8252L4.42259 12.7559L8 9.17851L11.5774 12.7559L11.6559 12.8252C11.9828 13.0794 12.4555 13.0563 12.7559 12.7559C13.0814 12.4305 13.0814 11.9028 12.7559 11.5774L9.17851 8L12.7559 4.42259C13.0814 4.09715 13.0814 3.56951 12.7559 3.24408C12.4555 2.94367 11.9828 2.92057 11.6559 3.17475L11.5774 3.24408L8 6.82149L4.42259 3.24408L4.34408 3.17475Z"
                fill="currentColor"
                fill-rule="evenodd"
              />
            </svg>
          </button>
        </span>
      </label>
      <div
        class="add_label"
      >
        multi-location-selector/add
      </div>
    </div>
    <input
      class="hidden_input"
    />
  </div>
</div>
`;
